pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go

```diff
@@ -639,3 +642,54 @@ func hugePagesInfo(vmi *v12.VirtualMachineInstance, domain *api.DomainSpec) (siz
 	}
 	return 0, "b", false, nil
 }
+
+func numaTune(vmi *v12.VirtualMachineInstance, domain *api.DomainSpec, topology *v1.Topology) error {
+	memoryMode := numaMemoryMode(vmi)
+	if memoryMode == "" {
+		return nil
+	}
+
+	if topology == nil || len(topology.NumaCells) == 0 {
+		// If there is no numa topology reported, we don't do anything.
+		return nil
+	}
+
+	cpumap := cpuToCell(topology)
+	numamap, err := involvedCells(cpumap, domain.CPUTune)
+	if err != nil {
+		return fmt.Errorf("failed to generate numa pinning information: %v", err)
+	}
+
+	var involvedCellIDs []string
+	for _, cell := range topology.NumaCells {
+		if _, exists := numamap[cell.Id]; exists {
+			involvedCellIDs = append(involvedCellIDs, strconv.Itoa(int(cell.Id)))
+		}
+	}
+
+	domain.NUMATune = &api.NUMATune{
+		Memory: api.NumaTuneMemory{
+			Mode:    memoryMode,
+			NodeSet: strings.Join(involvedCellIDs, ","),
+		},
+	}
+
+	return nil
+}
+
+func numaMemoryMode(vmi *v12.VirtualMachineInstance) string {
+	if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.NUMA != nil && vmi.Spec.Domain.CPU.NUMA.MemoryMode != nil {
+		switch *vmi.Spec.Domain.CPU.NUMA.MemoryMode {
+		case v12.MemoryModeStrict:
+			return "strict"
+		case v12.MemoryModePreferred:
+			return "preferred"
+		case v12.MemoryModeInterleave:
+			return "interleave"
+		default:
+			log.Log.Errorf("Unrecognized memoryMode '%s'", *vmi.Spec.Domain.CPU.NUMA.MemoryMode)
+			return ""
+		}
+	}
+	return ""
+}
```

Comment:
I'm not sure this is necessary. The whole beauty of creating another branded type is that you're guaranteed that the only possible values are the expected ones.

I'd change the calling code:
```go
memoryMode := numaMemoryMode(vmi)
```

to:
```go
if vmi.Spec.Domain.CPU == nil || vmi.Spec.Domain.CPU.NUMA == nil || vmi.Spec.Domain.CPU.NUMA.MemoryMode == nil {
    return fmt.Errorf("....")
}

memoryMode = *vmi.Spec.Domain.CPU.NUMA.MemoryMode
```

and remove this function altogether.
WDYT?

Reply:
I think when you are using `json` and custom types you can pass a value that is an underlying type, this is example in Go Playground:
https://go.dev/play/p/LR9LOlCF0MI

Reply:
but if we rely on validating webhooks

> And yes, we can rely on a validating webhooks running against every VMI.

then we can skip this `switch` because we are already validating allowed values :)

Reply:
Yes, for sure we can rely on webhooks :)
Weird that testing locally I can't just use any value like in your go playground example. I guess that we do have some kind of extra validation, I'm just not sure where it's coming from.

Either way, if you can rely on webhooks and delete this switch it'd be awesome :)

Reply:
removed